admittance_controller/include/admittance_controller/admittance_rule_impl.hpp

```diff
@@ -0,0 +1,534 @@
+    sensor_transform_ = get_transform(current_joint_state.positions, parameters_->ft_sensor_.frame_.id_,
+                                      parameters_->ft_sensor_.frame_.external_, success);
+    cog_transform_ = get_transform(current_joint_state.positions, parameters_->gravity_compensation_.frame_.id_,
+                                   parameters_->gravity_compensation_.frame_.external_, success);
+
+    // get all needed rotations
+    ee_rot_ = ee_transform_.block<3, 3>(0, 0);
+    control_rot_ = control_transform_.block<3, 3>(0, 0);
+    sensor_rot_ = sensor_transform_.block<3, 3>(0, 0);
+    cog_rot_ = cog_transform_.block<3, 3>(0, 0);
+    world_rot_ = world_transform_.block<3, 3>(0, 0);
+
+    // apply filter and update wrench_world_ vector
+    process_wrench_measurements(measured_wrench, world_rot_ * sensor_rot_, world_rot_ * cog_rot_);
+
+    // transform wrench_world_ into base frame
+    Eigen::Matrix<double, 3, 2> wrench_base = world_rot_.transpose() * wrench_world_;
+
+    // calculate desired cartesian velocity in control frame
+    Eigen::Matrix<double, 3, 2> desired_vel = convert_joint_deltas_to_cartesian_deltas(current_joint_state.positions,
+                                                                                       current_joint_state.velocities,
+                                                                                       parameters_->ft_sensor_.frame_.id_,
+                                                                                       success);
+    // set desired_vel to zero if not using feed forward input
+    desired_vel = desired_vel * parameters_->use_feedforward_commanded_input_;
+
+    // Compute admittance control law
+    calculate_admittance_rule(wrench_base, desired_vel, dt);
+
+    // calculate drift due to integrating the joint positions
+    Eigen::Matrix<double, 3, 1> admittance_position_base = reference_ft_transform_.block<3,1>(0, 3) + admittance_position_.block<3,1>(0, 3);
+    Eigen::Matrix<double, 3, 1> admittance_actually_position_base = ft_transform_.block<3,1>(0,3);
+    Eigen::Matrix<double, 3, 3> ft_rot = ft_transform_.block<3,3>(0,0);
+    Eigen::Matrix<double, 3, 3> reference_ft_rot = reference_ft_transform_.block<3,3>(0,0);
+    Eigen::Matrix<double, 3, 3> admittance_rot = admittance_position_.block<3,3>(0,0);
+    Eigen::Matrix<double, 3, 3> R = admittance_rot.transpose()*(reference_ft_rot.transpose()*reference_ft_rot);
+
+    // calculate correction term
+    Eigen::Matrix<double, 3, 2> correction_velocity;
+    correction_velocity.block<3, 1>(0, 0) = (admittance_position_base - admittance_actually_position_base);
+    Eigen::Vector3d V = get_rotation_axis(R);
+    double theta = acos(std::clamp((1.0 / 2.0) * (R.trace() - 1), -1.0, 1.0));
+    auto tmp = V[0] * (R(1, 2) - R(2, 1)) + V[1] * (R(2, 0) - R(0, 2))
+               + V[2] * (R(0, 1) - R(1, 0));
+    double sign = (tmp >= 0) ? 1.0 : -1.0;
+    correction_velocity.block<3, 1>(0, 1) = .1*sign*theta*V;
```

Comment:
This looks really hacky... very hard to read with those short names. Does it have to be so?

Reply:
No, it does not have to be. I will change the  names.